### ISSUE TITLE
Improve auto-completion on the Editor

### ISSUE THREAD

**Issue by spyder-issues-migrator** [spyder-ide/spyder#1732]
_From [hesiyuan...@gmail.com](https://code.google.com/u/111119374578212631612/) on 2014-02-20T02:02:59Z_

Spyder Version:  2.2.4
Python Version:  2.7.5
Qt Version    :  4.8.4, PyQt4 (API v2) 4.9.6 on Windows
pyflakes >=0.5.0:  0.7.3 (OK)
pep8 >=0.6      :  1.4.6 (OK)
IPython >=0.13  :  1.1.0 (OK)
rope >=0.9.2    :  0.9.4 (OK)
sphinx >=0.6.6  :  1.2b2 (OK)
matplotlib >=1.0:  1.3.0 (OK)
sympy >=0.7.0   :  0.7.3 (OK)
pylint >=0.25   :  1.0.0 (OK)

what i could assessment spyder ,is that it is a really good IDE for the python programmer . what i prefer is the object inspector .make the programing process more esay and happy . i really enjoy that . 
how ever , the spyer auto-compliment is very week . it only work when you enter a ful stop . that is really awful . especially you had define lots variant and function .
without the coding compliment , and code remind , the coding process spend lots of  time , and make coding not so much fun . 
hope the next version , the auto -compliment function could optimize .

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=1732_

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2014-02-20T05:33:35Z_

I don't understand exactly what you mean. Could you post a code example that shows the problem?

**Comment by spyder-issues-migrator** [comment]
_From [contrebasse](https://code.google.com/u/contrebasse/) on 2014-02-20T12:22:42Z_

I think that he means that the autocomplete feature work only after a dot (like after "np."). He would like autocomplete for all variables as soon as possible:

    long_variable_name = 10
    lo<autocomplete>

I had the inverse problem: I disabled autocomplete because it would autocomplete where it knows only one possibility and enter the name while I type.

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2014-02-20T18:14:25Z_

Oh I see, I think you are right. Then he's probably unaware the he can use \<Ctrl+Space> next to any name to get code completions.

About your problem, we could activate by default the option in

Preferences > Editor > Code Introspection/Analysis > Introspection > Show single completion

to avoid that unpleasant side effect. What do you think?

**Summary:** Weak auto-complete on the Editor   \<span class="oldvalue"> (was: spyder week auto compliment ) \</span>

**Comment by spyder-issues-migrator** [comment]
_From [hesiyuan...@gmail.com](https://code.google.com/u/111119374578212631612/) on 2014-02-20T19:13:00Z_

use \<Ctrl+Space> keyboard shortcut which had been occupied by the input software , to get code completion is good idea .it works . thanks
for the flesh programmer , i prefer the spyder could have such a feature like visual assistant  and  pydev :  
when you enter one or two letters , it will automatically  come out a relative hint box for you :(you could choose the object , or type it ,or new one )

timeeclipse=100
ti [ time ]
   [ timeeclipse  ]

although \<Ctrl+Space> shortcut auto completion could do this work , if there is no one tell me such a feature , i am still doing the tough letter typing work 
thanks again

**Comment by spyder-issues-migrator** [comment]
_From [contrebasse](https://code.google.com/u/contrebasse/) on 2014-02-21T10:52:45Z_

@-ccordoba12: I like that a single completion is filled with ctrl+space without showing the popup. The perfect behavior for me would be to always show the popup with the autocompletion (single or not) and autofill single completions when I hit ctrl+space. But adding yet another option for completion may clutter the config panel too much for this corner case. 

This problem aside I think that hesiyuan has a point. The autocompletion popup could show up after entering a dot (as it is now) OR after a (configurable) number of letters. I remember seeing this option in another IDE (I don't remember which one).
In the last case autocompletion wihout a popup could be annoying when creating new variables starting with the same letters.

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2014-02-24T04:48:44Z_

@-contrebasse: The best compromise then would be to kill the "Show single completion" option to *always* show the completion widget because with the newly introduced "smart tab key" you can hit \<Tab> to complete a single completion. What do you think?

I also think hesiyuan has a really good point here. I'm using Mathematica 9 a lot these days and it feels quite natural to see the completion widget after having written a couple of chars. I'll consider it for 2.3.1 because it doesn't seem that hard to implement.

**Summary:** Improve auto-completion on the Editor   \<span class="oldvalue"> (was: Weak auto-complete on the Editor ) \</span>  
  **Status:** Accepted  
  **Labels:** Cat-Editor MS-v2.3.1

**Comment by spyder-issues-migrator** [comment]
_From [contrebasse](https://code.google.com/u/contrebasse/) on 2014-03-03T01:23:56Z_

What is the difference between "smart tab key" and autocompletion ? It looks like the same thing to me.
Btw I usually use the tab key for indenting. In python it seems that I reindent my code as often as I use autocompeltion ^^

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2014-05-04T14:43:25Z_

@-contrebasse: Tab will indent when the context indicates that you want to do that (e.g. empty lines, beginning of line) but will complete when it seems appropriate (e.g. middle of a word).

By the way, I've been thinking a lot about this feature request and I'm afraid I'm going to say *no* to it.

Let me explain why: I want like all Spyder components behave more or less the same, so if we introduce this functionality, we will have to add it to the IPython console and later to the notebook. Since we don't have control over this products, there is almost nothing we can about it.

Sorry, but I think using Tab is a good compromise, besides being the standard key to get completions all IPython interfaces.

**Status:** WontFix  
  **Labels:** -MS-v2.3.1
